build-support/bin/generate_travis_yml.py

```diff
@@ -0,0 +1,771 @@
+from textwrap import dedent
+from typing import Dict, List, Optional
+
+import yaml
+
+
+HEADER = dedent("""\
+  # GENERATED, DO NOT EDIT!
+  # To change, edit `build-support/bin/generate_travis_yml.py` and run:
+  # ./pants --quiet run build-support/bin:generate_travis_yml > .travis.yml
+  """)
+
+# ----------------------------------------------------------------------
+# Stages
+# ----------------------------------------------------------------------
+
+class Stage(Enum):
+  bootstrap = "Bootstrap Pants"
+  bootstrap_cron = "Bootstrap Pants (Cron)"
+  test = "Test Pants"
+  test_cron = "Test Pants (Cron)"
+  build_stable = "Deploy Pants Pex"
+  build_unstable = "Deploy Pants Pex Unstable"
+
+  def condition(self) -> str:
+    is_cron = "type = cron"
+    is_not_cron = "type != cron"
+    return {
+      self.bootstrap: is_not_cron,
+      self.bootstrap_cron: is_cron,
+      self.test: is_not_cron,
+      self.test_cron: is_cron,
+      self.build_stable: r"tag IS present AND tag =~ ^release_.*$",
+      self.build_unstable: r"tag IS NOT present AND type NOT IN (pull_request, cron)"
+    }[self]
+
+  @classmethod
+  def all_entries(cls) -> List[Dict[str, str]]:
+    return [{"name": stage.value, "if": stage.condition()} for stage in Stage]
+
+# ----------------------------------------------------------------------
+# Env vars
+# ----------------------------------------------------------------------
+
+GLOBAL_ENV_VARS = [
+  'PANTS_CONFIG_FILES="${TRAVIS_BUILD_DIR}/pants.travis-ci.ini"',
+  'LC_ALL="en_US.UTF-8"',
+  'BOOTSTRAPPED_PEX_BUCKET=ci-public.pantsbuild.org',
+  'BOOTSTRAPPED_PEX_KEY_PREFIX=${TRAVIS_BUILD_NUMBER}/${TRAVIS_BUILD_ID}/pants.pex',
+  'BOOTSTRAPPED_PEX_URL_PREFIX=s3://${BOOTSTRAPPED_PEX_BUCKET}/${BOOTSTRAPPED_PEX_KEY_PREFIX}',
+  'PYENV_PY27_VERSION=2.7.15',
+  'PYENV_PY36_VERSION=3.6.8',
+  'PYENV_PY37_VERSION=3.7.2',
+  # NB: Linux shards use Pyenv to pre-install Python. We must not override PYENV_ROOT on those
+  # shards, or their Python will no longer work.
+  'PYENV_ROOT="${PYENV_ROOT:-${HOME}/.pants_pyenv}"',
+  'PATH="${PYENV_ROOT}/shims:${PATH}"',
+  'AWS_CLI_ROOT="${HOME}/.aws_cli"',
+  # NB: We use this verbose name so that AWS does not pick up the env var $AWS_ACCESS_KEY_ID on
+  # pull request builds. We only want this value to be populated on branch builds. Users of this
+  # env var (e.g. `deploy_to_s3.py`) are expected to re-export the env var as $AWS_ACCESS_KEY_ID.
+  'AWS_ACCESS_KEY_ID__TO_BE_REEXPORTED_ON_DEPLOYS=AKIAV6A6G7RQWPRUWIXR',
+  # This stores the encrypted AWS secret access key with the env var AWS_SECRET_ACCESS_KEY.
+  # Travis converts it back into its original decrypted value when ran in CI, per
+  # https://docs.travis-ci.com/user/environment-variables#defining-encrypted-variables-in-travisyml.
+  {"secure": "hFVAQGLVkexzTd3f9NF+JoG1dE+CPICKqOcdvQYv8+YB2rwwqu0/J6MnqKUZSmec4AM4ZvyPUBIHnSw8aMJysYs+GZ6iG/8ZRRmdbmo2WBPbSZ+ThRZxx/F6AjmovUmf8Zt366ZAZXpc9NHKREkTUGl6UL7FFe9+ouVnb90asdw="},
```

Comment:
The other elements of this list are strings -- it's not clear to me whether this dict object is encoded into a string somehow, or whether this would cause a runtime error.